File: PP/Lista2/replaceNth.scala
object replaceNth {
  def main(args : Array[String]) = {
    def replaceNth[A](xs : List[A], i : Int, x : A) : List[A] =
      xs match {
        case List() => List()
        case h :: t if i == 0 => x :: replaceNth(t, i - 1, x)
        case h :: t => h :: replaceNth(t, i - 1, x)
      }
  }
}

File: PP/Lista2/replaceNth.ml
(* O(n) *)
let rec repalceNth (xs, i, x) =
    match (xs, i, x) with
    | ([], _, _) -> []
    | ((h :: t), 0, x) -> x :: t
    | ((h :: t), i, x) -> h :: repalceNth (t, i - 1, x);;

let rec print_list xs =
    match xs with
    | [] -> ()
    | e::l -> print_char e ; print_string " " ; print_list l;;

print_list (repalceNth (['o';'l';'a';'m';'a';'k';'o';'t';'a'], 1, 's'));;
